Shopify/shopify_app
# Allow for `current_shopify_domain` to be nil with authenticated requests

![](https://media.tenor.com/M4iXZdH1UbEAAAAC/crying-sad.gif)

### What this PR does

There is a bug [with expired tokens](https://github.com/Shopify/shopify_app/issues/1536) that causes our CSP header to throw an error because `current_shopify_domain` isn't able to be retrieved because there is no current active session (it's expired).

Because `LoginProtection#current_shopify_domain` previously has been throwing and error when the domain couldn't be found from the session, this was throwing and exception before it could return a `401 - unauthorized` response.

#### `LoginProtection#current_shopify_domain`

This changes a few things that I'll want to make sure are acceptable design changes:

1. `current_shopify_domain` is allowed to be nil within `LoginProtection`
2. When `current_shopify_domain` is required for `302 - redirect`s in `LoginProtection` we will move responsibility for raising and error to those redirect calls.

#### Fetch requests don't evaluate to true with the `requirest.hr?` helper

The other problem preventing us from returning `401 - unauthorized` was that we were doing a full page redirect when the current session wasn't found. Our generated `home/index.html.erb` used the fetch API and `LoginProtection` was only looking for `xhr` requests to handle JS requests 🤦 . 

1. Added `requseted_by_javascript?` conditional to better determine API requests vs browser requests. [MDN indicates](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#textjavascript) the official MIME type is text/javascript though we'll support the legacy application/javascript as well.
2. Updated our `home_controller/index.html.erb` to add the correct content type header to act as a better example.

### Reviewer's guide to testing

1. Clone [test repo that reproduces expired token](https://github.com/dansundy/app_gem_error_test)
2. Add `Content-Type: "text/javascript"` header to `/fetch` request in `home_controller/index.html.erb`
3. Create test app with that test repo
4. Ensure server responds with expected `401` instead of `500` with bad tokens.

### Things to focus on

Is changing the raise behavior in `current_shopify_domain` acceptable? Or is this too big of a breaking change? If so, we can rescue in `FrameAncestors` and use our wildcard fallback domain. 

### Checklist

Before submitting the PR, please consider if any of the following are needed:
